tests/interfaces/test_batch_torch.py

```diff
@@ -0,0 +1,875 @@
+            mode="backward",
+            gradient_kwargs={"method": "adjoint_jacobian"},
+            interface="torch",
+        )[0]
+
+        assert dev.num_executions == 1
+        spy_execute.assert_called()
+        spy_gradients.assert_not_called()
+
+        res.backward()
+        spy_gradients.assert_called()
+
+
+class TestCaching:
+    """Test for caching behaviour"""
+
+    def test_cache_maxsize(self, mocker):
+        """Test the cachesize property of the cache"""
+        dev = qml.device("default.qubit", wires=1)
+        spy = mocker.spy(qml.interfaces.batch, "cache_execute")
+
+        def cost(a, cachesize):
+            with qml.tape.JacobianTape() as tape:
+                qml.RY(a[0], wires=0)
+                qml.RX(a[1], wires=0)
+                qml.probs(wires=0)
+
+            return execute(
+                [tape], dev, gradient_fn=param_shift, cachesize=cachesize, interface="torch"
+            )[0][0, 0]
+
+        params = torch.tensor([0.1, 0.2], requires_grad=True)
+        res = cost(params, cachesize=2)
+        res.backward()
+        cache = spy.call_args[0][1]
+
+        assert cache.maxsize == 2
+        assert cache.currsize == 2
+        assert len(cache) == 2
+
+    def test_custom_cache(self, mocker):
+        """Test the use of a custom cache object"""
+        dev = qml.device("default.qubit", wires=1)
+        spy = mocker.spy(qml.interfaces.batch, "cache_execute")
+
+        def cost(a, cache):
+            with qml.tape.JacobianTape() as tape:
+                qml.RY(a[0], wires=0)
+                qml.RX(a[1], wires=0)
+                qml.probs(wires=0)
+
+            return execute([tape], dev, gradient_fn=param_shift, cache=cache, interface="torch")[0][
+                0, 0
+            ]
+
+        custom_cache = {}
+        params = torch.tensor([0.1, 0.2], requires_grad=True)
+        res = cost(params, cache=custom_cache)
+        res.backward()
+
+        cache = spy.call_args[0][1]
+        assert cache is custom_cache
+
+    def test_caching_param_shift(self, tol):
+        """Test that, when using parameter-shift transform,
+        caching reduces the number of evaluations to their optimum."""
+        dev = qml.device("default.qubit", wires=1)
+
+        def cost(a, cache):
+            with qml.tape.JacobianTape() as tape:
+                qml.RY(a[0], wires=0)
+                qml.RX(a[1], wires=0)
+                qml.probs(wires=0)
+
+            return execute([tape], dev, gradient_fn=param_shift, cache=cache, interface="torch")[0][
+                0, 0
+            ]
+
+        # Without caching, 5 evaluations are required to compute
+        # the Jacobian: 1 (forward pass) + (2 shifts * 2 params)
+        params = torch.tensor([0.1, 0.2], requires_grad=True)
+        torch.autograd.functional.jacobian(lambda p: cost(p, cache=None), params)
+        assert dev.num_executions == 5
+
+        # With caching, 5 evaluations are required to compute

```

Comment:
The docstring for this test says that caching reduces the number of evaluations, but here the number of evaluations is the same in both cases.

Reply:
Yes 😕 For Autograd, this is true - without caching, the number of evals is 9. With caching, the number of evals is reduced to 5.

When I ported this test over to Torch, I discovered that... this is not the case for Torch! Torch seems to be smarter than autograd, and always does the optimum number of evals for the Jacobian (but not the Hessian).

I decided to keep this test anyway as a failsafe - let me modify the test docstring.